src/walks.jl

```diff
@@ -0,0 +1,159 @@
+
+function (::DefaultWalk)(recurse, x, ys...)
+  func, re = functor(x)
+  yfuncs = map(y -> functor(typeof(x), y)[1], ys)
+  re(map(recurse, func, yfuncs...))
+end
+
+"""
+    StructuralWalk()
+
+A structural variant of [`Functors.DefaultWalk`](@ref).
+The recursion behavior is identical, but the mapped children are not restructured.
+
+See [`fmapstructure`](@ref) for more information.
+"""
+struct StructuralWalk <: AbstractWalk end
+
+(::StructuralWalk)(recurse, x) = map(recurse, children(x))
+
+"""
+    ExcludeWalk(walk, fn, exclude)
+
+A walk that recurses nodes `(x, ys...)` according to `walk`,
+except when `exclude(x)` is true.
+Then, `fn(x, ys...)` is applied instead of recursing further.
+
+Typically wraps an existing `walk` for use with [`fmap`](@ref).
+"""
+struct ExcludeWalk{T, F, G} <: AbstractWalk
+  walk::T
+  fn::F
+  exclude::G
+end
+
+(walk::ExcludeWalk)(recurse, x, ys...) =
+  walk.exclude(x) ? walk.fn(x, ys...) : walk.walk(recurse, x, ys...)
+
+struct NoKeyword end
+
+usecache(::AbstractDict, x) = isleaf(x) ? anymutable(x) : ismutable(x)
+usecache(::Nothing, x) = false
+
+@generated function anymutable(x::T) where {T}
+  ismutabletype(T) && return true
+  subs =  [:(anymutable(getfield(x, $f))) for f in QuoteNode.(fieldnames(T))]
+  return Expr(:(||), subs...)
+end
+
+"""
+    CachedWalk(walk[; prune])
+
+A walk that recurses nodes `(x, ys...)` according to `walk` and storing the
+output of the recursion in a cache indexed by `x` (based on object ID).
+Whenever the cache already contains `x`, either:
+- `prune` is specified, then it is returned, or
+- `prune` is unspecified, and the previously cached recursion of `(x, ys...)`
+  returned.
+
+Typically wraps an existing `walk` for use with [`fmap`](@ref).
+"""
+struct CachedWalk{T, S} <: AbstractWalk
+  walk::T
+  prune::S
+  cache::IdDict{Any, Any}
+end
+CachedWalk(walk; prune = NoKeyword(), cache = IdDict()) =
+  CachedWalk(walk, prune, cache)
+
+function (walk::CachedWalk)(recurse, x, ys...)
+  should_cache = usecache(walk.cache, x)
+  if should_cache && haskey(walk.cache, x)
+    return walk.prune isa NoKeyword ? walk.cache[x] : walk.prune
+  else
+    ret = walk.walk(recurse, x, ys...)
+    if should_cache
+      walk.cache[x] = ret
+    end
+    return ret
+  end
+end
+
+"""
+    CollectWalk()
+
+A walk that recurses into a node `x` via [`Functors.children`](@ref),
+storing the recursion history in a cache.
+The resulting ordered recursion history is returned.
+
+See [`fcollect`](@ref) for more information.
+"""
+struct CollectWalk <: AbstractWalk
+  cache::Base.IdSet{Any}
+  output::Vector{Any}
+end
+CollectWalk() = CollectWalk(Base.IdSet(), Any[])
+
+# note: we don't have an `OrderedIdSet`, so we use an `IdSet` for the cache
+# (to ensure we get exactly 1 copy of each distinct array), and a usual `Vector`
+# for the results, to preserve traversal order (important downstream!).
+function (walk::CollectWalk)(recurse, x)
+  x in walk.cache && return walk.output
```

Comment:
Given this will be in 0.4, could we take the opportunity to clean up the caching behaviour? I was thinking to share logic with `usecache` so we're not accidentally identifying isbits types like SArrays as already visited.

Reply:
Since #39 didn't touch this, I didn't either, but I can make this apply the same caching rules.